PRDs/20251217-allowlist-resolver/PRD.md

```diff
@@ -0,0 +1,410 @@
+├── IAccessDeniedHandler.java           # Custom handler interface
+├── config/
+│   ├── AllowlistConfiguration.java     # Configuration POJO
+│   ├── AllowlistConfigurationLoader.java  # YAML loader
+│   ├── SchemePolicy.java               # Enum: ALLOW, DENY
+│   ├── FileSystemRule.java             # File path rules
+│   ├── HttpRule.java                   # Domain/path rules
+│   └── JarRule.java                    # JAR resource rules
+└── denylist/
+    ├── BuiltInDenylist.java            # Immutable security rules
+    ├── NetworkDenylist.java            # IP/hostname patterns
+    └── FileSystemDenylist.java         # Sensitive path patterns
+```
+
+### Integration Flow
+
+```
+User Request (URI)
+       │
+       ▼
+┌──────────────────┐
+│  IModuleLoader   │──────┐
+│  IDocumentLoader │      │
+│  IConstraintLoader      │
+│  XMLResolver     │      │
+└──────────────────┘      │
+       │                  │
+       ▼                  ▼
+┌──────────────────────────────────────┐
+│      IAllowlistUriResolver           │
+│  ┌────────────────────────────────┐  │
+│  │ 1. Check built-in denylist    │  │
+│  │ 2. Check scheme policy        │  │
+│  │ 3. Check resource-specific    │  │
+│  │    rules (file/http/jar)      │  │
+│  │ 4. Log attempt                │  │
+│  │ 5. Return URI or throw        │  │
+│  └────────────────────────────────┘  │
+└──────────────────────────────────────┘
+       │
+       ▼
+   Allowed URI → Resource Access
+       or
+   AccessDeniedException → Blocked
+```
+
+---
+
+## API Design
+
+### Programmatic Configuration
+
+```java
+// Strict server mode - HTTPS only
+AllowlistUriResolver serverResolver = AllowlistUriResolver.builder()
+    .denyAllSchemes()
+    .allowScheme("https")
+    .allowDomain("pages.nist.gov")
+    .allowDomain("raw.githubusercontent.com", PathPolicy.prefix("/metaschema-framework/"))
+    .onAccessDenied((uri, reason) -> {
+        auditLog.warn("Blocked resource access: {} - {}", uri, reason);
+        throw new AccessDeniedException(uri, reason);
+    })
+    .build();
+
+// Development mode - allow local files
+AllowlistUriResolver devResolver = AllowlistUriResolver.builder()
+    .allowScheme("file")
+    .allowScheme("https")
+    .allowScheme("jar")
+    .allowDirectory(Path.of("/workspace/schemas"), DirectoryScope.RECURSIVE)
+    .allowDirectory(Path.of("/workspace/examples"), DirectoryScope.RECURSIVE)
+    .build();
+
+// Hierarchical - inherit global with overrides
+AllowlistUriResolver.setGlobalDefaults(serverResolver);
+
+IModuleLoader loader = context.newModuleLoader();
+loader.setUriResolver(AllowlistUriResolver.builder()
+    .inheritGlobalDefaults()
+    .allowScheme("file")  // Override for this loader
+    .allowDirectory(trustedSchemaPath)
+    .build());
+```
+
+### YAML Configuration
+
+```yaml
+# allowlist-config.yaml
+schemes:
+  default: deny
+  allow:
+    - https
+    - jar
+  deny:
+    - file
+    - http
+    - ftp
```

Comment:
Can this be addressed using the scheme keys (e.g., `http`) below? Should we organize all schemes under the `schemes` key?

Reply:
The YAML structure has been refactored.